source/material_model/rheology/peierls_creep.cc

```diff
@@ -0,0 +1,169 @@
+/*
+  Copyright (C) 2020 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+#include <aspect/material_model/rheology/peierls_creep.h>
+#include <aspect/utilities.h>
+
+#include <deal.II/base/signaling_nan.h>
+#include <deal.II/base/parameter_handler.h>
+
+
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    namespace Rheology
+    {
+      template <int dim>
+      PeierlsCreep<dim>::PeierlsCreep ()
+      {}
+
+
+
+      template <int dim>
+      double
+      PeierlsCreep<dim>::compute_viscosity (const double strain_rate,
+                                            const double temperature,
+                                            const unsigned int composition) const
+      {
+        double viscosity = 0.0;
+
+        switch (peierls_creep_flow_law)
+          {
+            case mei_viscosity_approx:
+            {
+              /**
+               * An approximation of the Peierls creep formulation from Mei et al. (2010), where stress is
+               * replaced with strain rate (second invariant) when calculation viscosity. This substitution
+               * require two additional fitting parameters (g,q). Details of this derivation can be found
+               * in the following document: https://ucdavis.app.box.com/s/scaorcblr9u294836pgk4hyf60gv7psr.
+               * The formulation for the viscosity is:
+               *  ((0.5*g*sigma_p)/(A*((g*sigma_p)^2)^(1/(s+2)))) * exp((E/(R*T))*((1-g^p)/(s+2))) * edot_ii^(1/(s+2)-1)
+               * where
+               * s = (E/(R*T))*p*(g^p)
+               * A is the Peierls prefactor term (1/s 1/Pa^2)
+               * E is the Peierls activation energy (J/mol)
+               * edot_ii is the second invariant of the deviatoric strain rate
+               * T is temperature (K)
+               * R is the gas constant
+               */
+              const double s = (activation_energies_peierls[composition] / (constants::gas_constant * temperature)) *
+                               peierls_fitting_exponent * std::pow(peierls_fitting_parameter,peierls_fitting_exponent);
+
+              const double left_term = 0.5 * (peierls_fitting_parameter * peierls_stress) /
+                                       std::pow((prefactors_peierls[composition] * std::pow(peierls_fitting_parameter * peierls_stress,2)),( 1 / (s + 2)));
```

Comment:
looking at Magali's derivation, it doesn't look like $n$ needs to be a fixed exponent (the powers of two in the viscosity expression). The Raterron et al. (2004) expression for Peierls creep in olivine has n=0, and it would be nice to also use this derivation to approximate their expression.

Reply:
It might also be good to follow other material models in ASPECT and include an activation volume for Peierls creep ("Eact" -> "Eact + P*Vact")

Reply:
I'll go back and check if we can use a general value of `n`. @mibillen - Thoughts on both comments above?

Reply:
I'll look at this today. Originally, I only derived this for the specific values of n, q, and p for the Mei flow law.

Reply:
Certainly +P*Vact can be added without any problems. 
The derivation at the moment is only applicable to the case p = 1, but there might be a clever way to get round this.